File: SlimFileManager/src/main/java/com/slim/slimfilemanager/utils/file/BaseFile.java
package com.slim.slimfilemanager.utils.file;

import java.io.File;

public abstract class BaseFile {

    protected String path;

    public static BaseFile getBlankFile() {
        return new BaseFile() {
            @Override
            public boolean delete() {
                return false;
            }

            @Override
            public boolean exists() {
                return false;
            }

            @Override
            public String getName() {
                return "";
            }

            @Override
            public String getParent() {
                return "";
            }

            @Override
            public String getPath() {
                return "";
            }

            @Override
            public String getRealPath() {
                return "";
            }

            @Override
            public boolean isDirectory() {
                return false;
            }

            @Override
            public void getFile(GetFileCallback callback) {
            }

            @Override
            public long length() {
                return 0;
            }

            @Override
            public String[] list() {
                return new String[0];
            }

            @Override
            public String getExtension() {
                return "";
            }

            @Override
            public long lastModified() {
                return 0;
            }

            @Override
            public Object getRealFile() {
                return new Object();
            }
        };
    }

    public abstract boolean delete();

    public abstract boolean exists();

    public abstract String getName();

    public abstract String getParent();

    public abstract String getPath();

    public abstract String getRealPath();

    public abstract boolean isDirectory();

    public abstract void getFile(GetFileCallback callback);

    public abstract long length();

    public abstract String[] list();

    public abstract String getExtension();

    public abstract long lastModified();

    public abstract Object getRealFile();

    public boolean isFile() {
        return !isDirectory();
    }

    public interface GetFileCallback {
        void onGetFile(File file);
    }
}

File: manager/src/main/java/com/slim/slimfilemanager/utils/file/BaseFile.kt
package com.slim.slimfilemanager.utils.file

import java.io.File

abstract class BaseFile {

    protected var filePath: String? = null

    abstract val name: String

    abstract val parent: String

    abstract val realPath: String

    abstract val isDirectory: Boolean

    abstract val extension: String

    abstract val realFile: Any

    val isFile: Boolean
        get() = !isDirectory

    abstract fun delete(): Boolean

    abstract fun exists(): Boolean

    abstract fun getPath(): String

    abstract fun getFile(callback: GetFileCallback)

    fun getFile(callback: (File) -> Unit) {
        getFile(callback)
    }

    abstract fun length(): Long

    abstract fun list(): Array<String>

    abstract fun lastModified(): Long

    interface GetFileCallback {
        fun onGetFile(file: File)
    }

    companion object {

        val blankFile: BaseFile
            get() = object : BaseFile() {

                override val name: String
                    get() = ""

                override val parent: String
                    get() = ""

                override val realPath: String
                    get() = ""

                override val isDirectory: Boolean
                    get() = false

                override val extension: String
                    get() = ""

                override val realFile: Any
                    get() = Any()

                override fun delete(): Boolean {
                    return false
                }

                override fun exists(): Boolean {
                    return false
                }

                override fun getPath(): String {
                    return ""
                }

                override fun getFile(callback: GetFileCallback) {}

                override fun length(): Long {
                    return 0
                }

                override fun list(): Array<String> {
                    return emptyArray()
                }

                override fun lastModified(): Long {
                    return 0
                }
            }
    }
}
